### PR TITLE
LogInterceptor

### PR DESCRIPTION
**Pull request by avemike**
LogInterceptor and within it async waitFor function which repetitively calls predicate function until it returns true, or max iterations number / timeout will be reached

### CODE REVIEW

File: tests/integration/smoke.ts [hunk 1]
```diff
@@ -16,11 +17,15 @@ describe('Smoke test', () => {
...
 
+    const logInterceptor = new LogInterceptor(apiServerProcess.stdout);
+
     apiServerProcess
       .on('error', (err) => done(err))
       .on('message', (msg) => {
         if (msg === 'Server started' && !isApiStarted) {
           isApiStarted = true;
+          logInterceptor.expectLogs(1);
+          console.log(logInterceptor.getLogs());
```
Comment (kacepe) [inline tests/integration/smoke.ts:28]: Please remove console.log

File: tests/utils/LogInterceptor/index.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  getLogs() {
+    return this.logs;
+  }
+
+  cleanLogs() {
+    this.logs = [];
+  }
+
+  async waitFor(predicate: (logs: string[]) => Promise<boolean>, timeout: number, maxIterations: number) {
```
Comment (kacepe) [inline tests/utils/LogInterceptor/index.ts:20]: Can we have any tests for it?

Comment (kacepe) [reply tests/utils/LogInterceptor/index.ts:20]: In smoke for example

File: tests/utils/LogInterceptor/index.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    const finishTime = Date.now() + timeout;
+
+    while (Date.now() < finishTime && maxIterations--) {
```
Comment (Dziejo93) [inline tests/utils/LogInterceptor/index.ts:23]: imo `--` in if is misleading. I suggest to do it at the end of a while.

File: tests/utils/LogInterceptor/index.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  countErrors() {
+    return this.logs.filter(({ level }) => level === 50).length;
+  }
+
+  expectLogs(expectedCount: number, strict = true) {
+    const isExpectedCount = strict ? this.logs.length === expectedCount : this.logs.length >= expectedCount;
+
+    if (!isExpectedCount) {
+      throw new Error("Count of logs doesn't match expected count of logs");
```
Comment (Dziejo93) [inline tests/utils/LogInterceptor/index.ts:38]: "Count of logs => "Log count ? But I may be wrong :p

**Review by Dziejo93 — CHANGES_REQUESTED** [review]

File: tests/utils/LogInterceptor/index.ts (hunk 2 above)
Comment (Dziejo93) [reply tests/utils/LogInterceptor/index.ts:20]: maybe also unit test the class?

**Review by Dziejo93 — APPROVED** [review]

**Review by kacepe — APPROVED** [review]

**Pull request merged**